syntax = "proto3";

package university;

message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  string publisher = 5;
  int32 pageCount = 6;
  int32 stock = 7;
}

message Student {
  string id = 1;
  string name = 2;
  string studentNumber = 3;
  string email = 4;
  bool isActive = 5;
}

enum LoanStatus {
  ONGOING = 0;
  RETURNED = 1;
  LATE = 2;
}

message Loan {
  string id = 1;
  string studentId = 2;
  string bookId = 3;
  string loanDate = 4;
  string returnDate = 5;
  LoanStatus status = 6;
}

message IdRequest {
  string id = 1;
}

message Empty {}

service BookService {
  rpc ListBooks(Empty) returns (stream Book);
  rpc GetBook(IdRequest) returns (Book);
  rpc AddBook(Book) returns (Book);
  rpc UpdateBook(Book) returns (Book);
  rpc DeleteBook(IdRequest) returns (Empty);
}

service StudentService {
  rpc ListStudents(Empty) returns (stream Student);
  rpc GetStudent(IdRequest) returns (Student);
  rpc AddStudent(Student) returns (Student);
  rpc UpdateStudent(Student) returns (Student);
  rpc DeleteStudent(IdRequest) returns (Empty);
}

service LoanService {
  rpc ListLoans(Empty) returns (stream Loan);
  rpc GetLoan(IdRequest) returns (Loan);
  rpc AddLoan(Loan) returns (Loan);
  rpc ReturnLoan(IdRequest) returns (Loan);
}
